File: SRC.NET/LabelData.vb
Option Strict Off
Option Explicit On
Friend Class LabelData
	
	' Copyright (C) 1997-2012 Kei Sakamoto / Inui Tetsuyuki
	'Invalid_string_refer_to_original_code
	'Invalid_string_refer_to_original_code
	'Invalid_string_refer_to_original_code
	
	'繧､繝吶Φ繝医ョ繝ｼ繧ｿ縺ｮ繝ｩ繝吶Ν縺ｮ繧ｯ繝ｩ繧ｹ
	
	'Invalid_string_refer_to_original_code
	Public Name As Event.LabelType
	'陦檎分蜿ｷ
	Public LineNum As Integer
	'Invalid_string_refer_to_original_code
	Public Enable As Boolean
	'Invalid_string_refer_to_original_code
	Public AsterNum As Short
	
	'Invalid_string_refer_to_original_code
	Private StrData As String
	'繝ｩ繝吶Ν縺ｮ蛟区焚
	Private intParaNum As Short
	'Invalid_string_refer_to_original_code
	Private strParas() As String
	'Invalid_string_refer_to_original_code
	Private blnConst() As Boolean
	
	'繝代Λ繝｡繝ｼ繧ｿ縺ｮ蛟区焚
	Public Function CountPara() As Short
		CountPara = intParaNum
	End Function
	
	'繝ｩ繝吶Ν縺ｮ idx 逡ｪ逶ｮ縺ｮ繝代Λ繝｡繝ｼ繧ｿ
	Public Function Para(ByVal idx As Short) As String
		If idx <= intParaNum Then
			If blnConst(idx) Then
				Para = strParas(idx)
			Else
				Para = GetValueAsString(strParas(idx), True)
			End If
		End If
	End Function
	
	'Invalid_string_refer_to_original_code
	
	'Invalid_string_refer_to_original_code
	Public Property Data() As String
		Get
			DataControl = StrData
		End Get
		Set(ByVal Value As String)
			Dim i As Integer
			Dim lname As String
			
			'Invalid_string_refer_to_original_code
			StrData = Value
			
			'Invalid_string_refer_to_original_code
			lname = ListIndex(Value, 1)
			'Invalid_string_refer_to_original_code
			Select Case Asc(lname)
				Case 42 '*
					lname = Mid(lname, 2)
					Select Case Asc(lname)
						Case 42 '*
							lname = Mid(lname, 2)
							AsterNum = 3
						Case 45 '-
							lname = Mid(lname, 2)
							AsterNum = 2
						Case Else
							AsterNum = 2
					End Select
				Case 45 '-
					lname = Mid(lname, 2)
					Select Case Asc(lname)
						Case 42 '*
							lname = Mid(lname, 2)
							AsterNum = 1
						Case 45 '-
							lname = Mid(lname, 2)
					End Select
			End Select
			Select Case lname
				Case "繝励Ο繝ｭ繝ｼ繧ｰ"
					Name = Event_Renamed.LabelType.PrologueEventLabel
				Case "Invalid_string_refer_to_original_code"
					Name = Event_Renamed.LabelType.StartEventLabel
				Case "繧ｨ繝斐Ο繝ｼ繧ｰ"
					Name = Event_Renamed.LabelType.EpilogueEventLabel
				Case "繧ｿ繝ｼ繝ｳ"
					Name = Event_Renamed.LabelType.TurnEventLabel
				Case "Invalid_string_refer_to_original_code"
					Name = Event_Renamed.LabelType.DamageEventLabel
				Case "Invalid_string_refer_to_original_code"
					Name = Event_Renamed.LabelType.DestructionEventLabel
				Case "Invalid_string_refer_to_original_code"
					Name = Event_Renamed.LabelType.TotalDestructionEventLabel
				Case "Invalid_string_refer_to_original_code"
					Name = Event_Renamed.LabelType.AttackEventLabel
				Case "Invalid_string_refer_to_original_code"
					Name = Event_Renamed.LabelType.AfterAttackEventLabel
				Case "莨夊ｩｱ"
					Name = Event_Renamed.LabelType.TalkEventLabel
				Case "謗･隗ｦ"
					Name = Event_Renamed.LabelType.ContactEventLabel
				Case "騾ｲ蜈･"
					Name = Event_Renamed.LabelType.EnterEventLabel
				Case "閼ｱ蜃ｺ"
					Name = Event_Renamed.LabelType.EscapeEventLabel
				Case "Invalid_string_refer_to_original_code"
					Name = Event_Renamed.LabelType.LandEventLabel
				Case "菴ｿ逕ｨ"
					Name = Event_Renamed.LabelType.UseEventLabel
				Case "Invalid_string_refer_to_original_code"
					Name = Event_Renamed.LabelType.AfterUseEventLabel
				Case "螟牙ｽ｢"
					Name = Event_Renamed.LabelType.TransformEventLabel
				Case "Invalid_string_refer_to_original_code"
					Name = Event_Renamed.LabelType.CombineEventLabel
				Case "Invalid_string_refer_to_original_code"
					Name = Event_Renamed.LabelType.SplitEventLabel
				Case "Invalid_string_refer_to_original_code"
					Name = Event_Renamed.LabelType.FinishEventLabel
				Case "Invalid_string_refer_to_original_code"
					Name = Event_Renamed.LabelType.LevelUpEventLabel
				Case "蜍晏茜譚｡莉ｶ"
					Name = Event_Renamed.LabelType.RequirementEventLabel
				Case "蜀埼幕"
					Name = Event_Renamed.LabelType.ResumeEventLabel
				Case "Invalid_string_refer_to_original_code"
					Name = Event_Renamed.LabelType.MapCommandEventLabel
				Case "Invalid_string_refer_to_original_code"
					Name = Event_Renamed.LabelType.UnitCommandEventLabel
				Case "Invalid_string_refer_to_original_code"
					Name = Event_Renamed.LabelType.EffectEventLabel
				Case Else
					Name = Event_Renamed.LabelType.NormalLabel
			End Select
			
			'繝代Λ繝｡繝ｼ繧ｿ
			intParaNum = ListLength(Value)
			If intParaNum = -1 Then
				DisplayEventErrorMessage(CurrentLineNum, "Invalid_string_refer_to_original_code")
				Exit Property
			End If
			ReDim strParas(intParaNum)
			ReDim blnConst(intParaNum)
			For i = 2 To intParaNum
				strParas(i) = ListIndex(Value, i)
				'Invalid_string_refer_to_original_code
				If IsNumeric(strParas(i)) Then
					blnConst(i) = True
				ElseIf PDList.IsDefined(strParas(i)) Then 
					If InStr(strParas(i), "荳ｻ莠ｺ蜈ｬ") <> 1 And InStr(strParas(i), "繝偵Ο繧､繝ｳ") <> 1 Then
						blnConst(i) = True
					End If
				ElseIf UDList.IsDefined(strParas(i)) Then 
					blnConst(i) = True
				ElseIf IDList.IsDefined(strParas(i)) Then 
					blnConst(i) = True
				Else
					Select Case strParas(i)
						Case "蜻ｳ譁ｹ", "Invalid_string_refer_to_original_code", "謨ｵ", "Invalid_string_refer_to_original_code"
							'Invalid_string_refer_to_original_code
							'UPGRADE_ISSUE: 前の行を解析できませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="82EBB1AE-1FCB-4FEF-9E6C-8736A316F8A7"' をクリックしてください。
							blnConst(i) = True
						Case "N", "W", "S", "E"
							If Name = Event_Renamed.LabelType.EscapeEventLabel Then
								blnConst(i) = True
							End If
						Case Else
							If Left(strParas(i), 1) = """" And Right(strParas(i), 1) = """" Then
								If InStr(strParas(i), "$(") = 0 Then
									strParas(i) = Mid(strParas(i), 2, Len(strParas(i)) - 2)
									blnConst(i) = True
								End If
							End If
					End Select
				End If
			Next 
		End Set
	End Property
End Class

File: SRC.Sharp/SRCCore/Events/LabelData.cs
// Copyright (C) 1997-2012 Kei Sakamoto / Inui Tetsuyuki
// 本プログラムはフリーソフトであり、無保証です。
// 本プログラムはGNU General Public License(Ver.3またはそれ以降)が定める条件の下で
// 再頒布または改変することができます。
using SRCCore.Exceptions;
using SRCCore.Lib;
using SRCCore.VB;

namespace SRCCore.Events
{
    // イベントデータのラベルのクラス
    public class LabelData
    {
        // ラベル名
        public LabelType Name;
        // 行番号
        public int EventDataId;
        // ラベルが有効か？
        public bool Enable;
        // アスタリスクの指定状況
        public int AsterNum;

        // ラベル全体
        private string StrData;
        // ラベルの個数
        private int intParaNum;
        // ラベルの各パラメータ
        private string[] strParas;
        // パラメータが固定値？
        private bool[] blnConst;

        protected SRC SRC { get; }
        private Expressions.Expression Expression => SRC.Expression;

        public LabelData(SRC src)
        {
            SRC = src;
        }

        public override string ToString()
        {
            return $"{EventDataId}: {StrData}";
        }

        // パラメータの個数
        public int CountPara()
        {
            return intParaNum;
        }

        // ラベルの idx 番目のパラメータ
        public string Para(int idx)
        {
            string ParaRet = default;
            if (idx <= intParaNum)
            {
                if (blnConst[idx])
                {
                    ParaRet = strParas[idx];
                }
                else
                {
                    ParaRet = Expression.GetValueAsString(strParas[idx], true);
                }
            }

            return ParaRet;
        }

        // ラベル全体を取り出す
        // ラベル全体を設定
        public string Data
        {
            get
            {
                return StrData;
            }
            set
            {
                // ラベル全体
                StrData = value;

                //        TODO Impl
                int i;
                string lname;

                // ラベル名
                lname = GeneralLib.ListIndex(value, 1);
                // 「*」は省く
                switch (Strings.Asc(lname))
                {
                    case 42: // *
                        {
                            lname = Strings.Mid(lname, 2);
                            switch (Strings.Asc(lname))
                            {
                                case 42: // *
                                    {
                                        lname = Strings.Mid(lname, 2);
                                        AsterNum = 3;
                                        break;
                                    }

                                case 45: // -
                                    {
                                        lname = Strings.Mid(lname, 2);
                                        AsterNum = 2;
                                        break;
                                    }

                                default:
                                    {
                                        AsterNum = 2;
                                        break;
                                    }
                            }

                            break;
                        }

                    case 45: // -
                        {
                            lname = Strings.Mid(lname, 2);
                            switch (Strings.Asc(lname))
                            {
                                case 42: // *
                                    {
                                        lname = Strings.Mid(lname, 2);
                                        AsterNum = 1;
                                        break;
                                    }

                                case 45: // -
                                    {
                                        lname = Strings.Mid(lname, 2);
                                        break;
                                    }
                            }

                            break;
                        }
                }

                switch (lname ?? "")
                {
                    case "プロローグ":
                        {
                            Name = Events.LabelType.PrologueEventLabel;
                            break;
                        }

                    case "スタート":
                        {
                            Name = Events.LabelType.StartEventLabel;
                            break;
                        }

                    case "エピローグ":
                        {
                            Name = Events.LabelType.EpilogueEventLabel;
                            break;
                        }

                    case "ターン":
                        {
                            Name = Events.LabelType.TurnEventLabel;
                            break;
                        }

                    case "損傷率":
                        {
                            Name = Events.LabelType.DamageEventLabel;
                            break;
                        }

                    case "破壊":
                        {
                            Name = Events.LabelType.DestructionEventLabel;
                            break;
                        }

                    case "全滅":
                        {
                            Name = Events.LabelType.TotalDestructionEventLabel;
                            break;
                        }

                    case "攻撃":
                        {
                            Name = Events.LabelType.AttackEventLabel;
                            break;
                        }

                    case "攻撃後":
                        {
                            Name = Events.LabelType.AfterAttackEventLabel;
                            break;
                        }

                    case "会話":
                        {
                            Name = Events.LabelType.TalkEventLabel;
                            break;
                        }

                    case "接触":
                        {
                            Name = Events.LabelType.ContactEventLabel;
                            break;
                        }

                    case "進入":
                        {
                            Name = Events.LabelType.EnterEventLabel;
                            break;
                        }

                    case "脱出":
                        {
                            Name = Events.LabelType.EscapeEventLabel;
                            break;
                        }

                    case "収納":
                        {
                            Name = Events.LabelType.LandEventLabel;
                            break;
                        }

                    case "使用":
                        {
                            Name = Events.LabelType.UseEventLabel;
                            break;
                        }

                    case "使用後":
                        {
                            Name = Events.LabelType.AfterUseEventLabel;
                            break;
                        }

                    case "変形":
                        {
                            Name = Events.LabelType.TransformEventLabel;
                            break;
                        }

                    case "合体":
                        {
                            Name = Events.LabelType.CombineEventLabel;
                            break;
                        }

                    case "分離":
                        {
                            Name = Events.LabelType.SplitEventLabel;
                            break;
                        }

                    case "行動終了":
                        {
                            Name = Events.LabelType.FinishEventLabel;
                            break;
                        }

                    case "レベルアップ":
                        {
                            Name = Events.LabelType.LevelUpEventLabel;
                            break;
                        }

                    case "勝利条件":
                        {
                            Name = Events.LabelType.RequirementEventLabel;
                            break;
                        }

                    case "再開":
                        {
                            Name = Events.LabelType.ResumeEventLabel;
                            break;
                        }

                    case "マップコマンド":
                        {
                            Name = Events.LabelType.MapCommandEventLabel;
                            break;
                        }

                    case "ユニットコマンド":
                        {
                            Name = Events.LabelType.UnitCommandEventLabel;
                            break;
                        }

                    case "特殊効果":
                        {
                            Name = Events.LabelType.EffectEventLabel;
                            break;
                        }

                    default:
                        {
                            Name = Events.LabelType.NormalLabel;
                            break;
                        }
                }

                // パラメータ
                intParaNum = GeneralLib.ListLength(value);
                if (intParaNum == -1)
                {
                    // TODO Impl
                    throw new TerminateException("ラベルの引数の括弧の対応が取れていません");
                    //Event.DisplayEventErrorMessage(Event.CurrentLineNum, "ラベルの引数の括弧の対応が取れていません");
                    //return;
                }

                strParas = new string[(intParaNum + 1)];
                blnConst = new bool[(intParaNum + 1)];
                var loopTo = intParaNum;
                for (i = 2; i <= loopTo; i++)
                {
                    strParas[i] = GeneralLib.ListIndex(value, i);
                    // パラメータが固定値かどうか判定
                    if (Information.IsNumeric(strParas[i]))
                    {
                        blnConst[i] = true;
                    }
                    else if (SRC.PDList.IsDefined(strParas[i]))
                    {
                        if (Strings.InStr(strParas[i], "主人公") != 1 && Strings.InStr(strParas[i], "ヒロイン") != 1)
                        {
                            blnConst[i] = true;
                        }
                    }
                    else if (SRC.UDList.IsDefined(strParas[i]))
                    {
                        blnConst[i] = true;
                    }
                    else if (SRC.IDList.IsDefined(strParas[i]))
                    {
                        blnConst[i] = true;
                    }
                    else
                    {
                        switch (strParas[i] ?? "")
                        {
                            case "味方":
                            case "ＮＰＣ":
                            case "敵":
                            case "中立":
                            case "全":
                                blnConst[i] = true;
                                break;

                            case "N":
                            case "W":
                            case "S":
                            case "E":
                                if (Name == Events.LabelType.EscapeEventLabel)
                                {
                                    blnConst[i] = true;
                                }

                                break;

                            default:
                                if (Strings.Left(strParas[i], 1) == "\"" && Strings.Right(strParas[i], 1) == "\"")
                                {
                                    if (Strings.InStr(strParas[i], "$(") == 0)
                                    {
                                        strParas[i] = Strings.Mid(strParas[i], 2, Strings.Len(strParas[i]) - 2);
                                        blnConst[i] = true;
                                    }
                                }

                                break;
                        }
                    }
                }
            }
        }
    }
}
